test/e2e/framework/create.go

```diff
@@ -0,0 +1,528 @@
+	// populating and patching it. It returns an error or
+	// a cleanup function for the created entity.
+	Create(f *Framework) (func() error, error)
+}
+
+var Factories = map[What]ItemFactory{
+	{"ClusterRole"}:        &ClusterRoleFactory{},
+	{"ClusterRoleBinding"}: &ClusterRoleBindingFactory{},
+	{"DaemonSet"}:          &DaemonSetFactory{},
+	{"Role"}:               &RoleFactory{},
+	{"RoleBinding"}:        &RoleBindingFactory{},
+	{"ServiceAccount"}:     &ServiceAccountFactory{},
+	{"StatefulSet"}:        &StatefulSetFactory{},
+	{"StorageClass"}:       &StorageClassFactory{},
+	{"Service"}:            &ServiceFactory{},
+}
+
+func visitManifests(cb func(data []byte) error, files ...string) error {
+	for _, fileName := range files {
+		Logf("parsing %s", fileName)
+		// Ultimately this should use the "testfiles" package from https://github.com/kubernetes/kubernetes/pull/69105
+		// For now we just approximate it.
+		data, err := testfiles.Read(fileName)
+		if err != nil {
+			Failf("reading manifest file: %v", err)
+		}
+
+		// Split at the "---" separator before working on
+		// individual item. Only works for .yaml.
+		//
+		// We need to split ourselves because we need access
+		// to each original chunk of data for
+		// runtime.DecodeInto. kubectl has its own
+		// infrastructure for this, but that is a lot of code
+		// with many dependencies.
+		items := bytes.Split(data, []byte("\n---"))
+
+		for _, item := range items {
+			if err := cb(item); err != nil {
+				return errors.Wrap(err, fileName)
+			}
+		}
+	}
+	return nil
+}
+
+// PatchName makes the name of some entity unique by appending the
+// generated unique name.
+func (f *Framework) PatchName(item *string) {
+	if *item != "" {
+		*item = *item + "-" + f.UniqueName
+	}
+}
+
+// PatchNamespace moves the entity into the test's namespace.  Not
+// all entities can be namespaced. For those, the name also needs to be
+// patched.
+func (f *Framework) PatchNamespace(item *string) {
+	if f.Namespace != nil {
+		*item = f.Namespace.GetName()
+	}
+}
+
+// PatchItem recursively fixes the name and namespace of an entity
+// such that each test gets its own private copy of the entity which
+// in turn uses the private copies of the other entities created by
+// the test.
+func (f *Framework) PatchItem(item interface{}) {
+	Logf("patching original content of %T:\n%s", item, PrettyPrint(item))
+	f.patchItemRecursively(item)
+}
+
+func (f *Framework) patchItemRecursively(item interface{}) {
+	switch item := item.(type) {
+	case *rbac.Subject:
+		f.PatchNamespace(&item.Namespace)
+	case *rbac.RoleRef:
+		// TODO: avoid hard-coding this special name. Perhaps add a Framework.PredefinedRoles
+		// which contains all role names that are defined cluster-wide before the test starts?
+		// All those names are excempt from renaming. That list could be populated by querying
+		// and get extended by tests.
+		if item.Name != "e2e-test-privileged-psp" {
```

Comment:
Or create one per driver?

Reply:
I don't know how `e2e-test-privileged-psp` gets set up. It's not done by the `e2e.test` binary at the moment. My understanding was that it is something that has to be configured in the cluster before starting the test.